### ISSUE TITLE
Recursive schemas lead to infinite recursion

### ISSUE THREAD

**Issue by jonmast** [anatine/zod-plugins#123]
With the recursive JSON schema from the zod docs:
```
const literalSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
type Literal = z.infer<typeof literalSchema>;
type Json = Literal | { [key: string]: Json } | Json[];
const jsonSchema: z.ZodType<Json> = z.lazy(() =>
  z.union([literalSchema, z.array(jsonSchema), z.record(jsonSchema)])
);
`

Sometimes generating mock JSON works, but often it will infinitely recurse until the stack overflows. It would be nice to have some sort of detection for this scenario and limit the recursion to a reasonable depth.

**Comment by Brian-McBride** [comment]
Yeah, the parser just walks through the tree. I guess maybe working with the `z.lazy` in some way? 
It makes me wonder if the mock generator could be rewritten using Zod's own validator parser codebase slightly modified to inject the mock values instead of parsing an actual object? Then it would align with how Zod works. 

Currently, it just walks through the zod schema object. What is the best way to stop infinite recursion in that way? Passing a counter into the object or array parser?

**Comment by Brian-McBride** [comment]
I was looking to build something here. I suppose each one of the worker functions could take a recursion variable and in the case of parsing objects in objects ticking it up by one. But that's not real recursion. It seems that this probably only happens in the case of Lazy. So maybe that is the time to tick up recursion? But I'm not sure if the lazy type only shows up once or if it is in very loop. 

I don't have time to dig in, but I see your error. I you just change your final scheme to `z.laz( () => z.record(jsonSchema) )` you'll end up with out of memory every time.
